net: Add support for Intel 82599 (ixgbe) NIC & performed performance optimization.

KuangjuX opened the pull request:
This is a relatively large PR that mainly does the following things:
- Referencing [ixy.rs](https://github.com/ixy-languages/ixy.rs), implemented the Intel 82599 (ixgbe) driver and optimized its performance to a certain extent.
- Completed the detection and initialization of ixgbe in `axdriver`.
- Added a new platform `x86-pc-oslab` for adapting to physical machines and conducting real network card development and testing.
- Based on #99, re-implemented the ixgbe driver and protocol stack adaptation for the network interface.
- Currently, the performance tested using ab (Apache HTTP server benchmarking tool) can reach 25000+ requests/second. TCP packet receiving performance tested with `iperf` can reach 3.5G, while packet sending performance can reach 4.0G. UDP packet receiving performance tested can reach 3.4G (not accurately measured), and packet sending performance reaches 9+G. Using `bw_client` for testing, the packet receiving performance reaches at least 6G (not accurately measured, as the client did not fully send), and packet sending can fully saturate at 10G.
![WechatIMG1507](https://github.com/rcore-os/arceos/assets/56680481/f2b76437-eb7e-429d-a6ec-47f0cac0d5d1)
![WechatIMG1506 1](https://github.com/rcore-os/arceos/assets/56680481/c56b0944-ebfe-442c-a3a6-a0fa9dfbf271)
- Still, there are some related tasks that need to be done, such as supporting network card interrupts and enabling multiple queues and RSS in multi-core. These will be submitted in subsequent PRs.

KuangjuX commented:
remove submodule, request review @equation314

equation314 commented on crates/driver_net/Cargo.toml:
```diff
@@ -9,6 +9,12 @@ homepage = "https://github.com/rcore-os/arceos"
 spin = "0.9"
+log = "0.4"
 driver_common = { path = "../driver_common" }
+ixgbe-driver = {git = "https://github.com/KuangjuX/ixgbe-driver.git", rev = "1a93244"}
```
`optional = true`

equation314 commented on crates/driver_net/src/lib.rs:
```diff
@@ -3,11 +3,19 @@
+#[macro_use]
+extern crate log;
+
+mod ixgbe;
```
`#[cfg(feature = "ixgbe")]`

equation314 replied:
pub mod ixgbe

equation314 commented on modules/axconfig/Cargo.toml:
Remove all `platform-pc-x86-2` features, keep the toml file only.

equation314 commented on crates/driver_net/src/lib.rs:
```diff
@@ -3,11 +3,19 @@
 use core::ptr::NonNull;
 
+#[cfg(feature = "ixgbe")]
+pub use ixgbe::{IxgbeHal, IxgbeNic, PhysAddr as IxgbePhysAddr, INTEL_82599, INTEL_VEND};
```
Exported by `pub mod ixgbe`

equation314 commented on modules/axdriver/src/drivers.rs:
```diff
@@ -66,3 +66,51 @@ cfg_if::cfg_if! {
+                        let bar_info = root.bar_info(bdf, 0).unwrap();
+                        match bar_info {
+                            driver_pci::BarInfo::Memory {
+                                address_type: _addr_type,
```
Use `..` to omit unused fields.

equation314 commented on modules/axdriver/src/ixgbe.rs:
```diff
@@ -0,0 +1,41 @@
+use core::ptr::NonNull;
+use driver_net::{IxgbeHal, IxgbePhysAddr};
+
+pub struct IxgbehalImpl;
```
`IxgbeHalImpl`

equation314 commented on modules/axfs/Cargo.toml:
```diff
@@ -17,6 +17,8 @@ myfs = ["dep:crate_interface"]
 
 default = ["devfs", "ramfs", "fatfs"]
+# default = ["devfs", "ramfs", "fatfs", "use-ramdisk"] # ramdisk
+
```
Remove this line.

equation314 commented on modules/axruntime/Cargo.toml:
Remove commented lines.

equation314 commented on Cargo.lock:
```diff
@@ -1421,6 +1434,21 @@ dependencies = [
 ]
 
+[[package]]
+name = "smoltcp"
```
Why are there two smoltcp packages?

KuangjuX replied:
Sorry, due to the inclusion of `smoltcp` in `ixgbe-driver` for debugging print information, it led to different versions of `smoltcp`. I will replace it with the same version of `smoltcp`.

equation314 commented on modules/axdriver/src/drivers.rs:
```diff
@@ -66,3 +66,51 @@ cfg_if::cfg_if! {
+
+                        // Initialize the device
+                        // These can be changed according to the requirments specified in the ixgbe init function.
+                        const QS: u16 = 1;
```
Why queue size is only 1?

KuangjuX replied:
Sorry, this is my mistake. Here, `QS` actually means the number of network card queues, which may be ambiguous.

equation314 commented on crates/driver_net/src/ixgbe.rs:
```diff
@@ -0,0 +1,157 @@
+    }
+
+    fn recycle_rx_buffer(&mut self, rx_buf: NetBufPtr) -> DevResult {
+        let rx_buf =
```
Add a method to convert `NetBufPtr` to `IxgbeNetBuf`

KuangjuX replied:
Since `IxgbeNetBuf` is defined in the `ixgbe-driver` crate and `NetBufPtr` is a generic data structure, implementing it as member methods of both seems inappropriate. Perhaps it could be implemented as external non-member functions?

KuangjuX replied:
It seems that implementing `From` and `Into` for `NetBufPtr` is possible.

equation314 commented on crates/driver_net/src/ixgbe.rs:
```diff
@@ -0,0 +1,157 @@
+            Ok(self.rx_buffer_queue.pop_front().unwrap())
+        } else {
+            // RX queue is empty, receive from ixgbe NIC.
+            match self.inner.receive_packets(0, RECV_BATCH_SIZE, |rx_buf| {
```
Add a method to convert `IxgbeNetBuf ` to `NetBufPtr`

equation314 commented on platforms/x86_64-pc-oslab.toml:
```diff
@@ -1,7 +1,7 @@
 # Platform identifier.
-platform = "pc-x86-2"
+platform = "x86_64-pc-oslab"
 
```
Add `family` field

equation314 commented on modules/axfs/Cargo.toml:
```diff
@@ -18,6 +18,7 @@ use-ramdisk = []
 
 default = ["devfs", "ramfs", "fatfs"]
 
+
```
Remove the blank line.

equation314 commented on modules/axhal/Cargo.toml:
```diff
@@ -15,6 +15,7 @@ fp_simd = []
 paging = ["axalloc", "page_table"]
 irq = []
 
+
```
Remove the blank line.

equation314 commented on doc/ixgbe.md:
````diff
@@ -0,0 +1,37 @@
+In `axfs/Cargo.toml`, you need to use `use_ramdisk` feature by default:
+
+```toml
+default = ["devfs", "ramfs", "fatfs", "use_ramdisk"]
````
`use-ramdisk` can be enabled in `libax`, these two changes is needn't.

equation314 commented on crates/driver_net/src/ixgbe.rs:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    fn rx_queue_size(&self) -> usize {
+        self.inner.num_rx_queues() as usize
```
`QS`

equation314 commented on crates/driver_net/src/ixgbe.rs:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    fn can_receive(&self) -> bool {
+        self.inner.can_receive(0).unwrap() || !self.rx_buffer_queue.is_empty()
```
Put `!self.rx_buffer_queue.is_empty()` before `self.inner.can_receive(0)`

equation314 commented on crates/driver_net/src/ixgbe.rs:
```diff
@@ -0,0 +1,161 @@
+        let tx_buf = IxgbeNetBuf::alloc(&self.mem_pool, size).map_err(|_| DevError::NoMemory)?;
+        // Use `ManuallyDrop` to avoid drop `tx_buf`.
+        let tx_buf = ManuallyDrop::new(tx_buf);
+        // In ixgbe, `raw_ptr` is the pool entry, `buf_ptr` is the packet ptr, `len` is packet len
```
Move this comment into the covert function

equation314 commented on crates/driver_net/src/ixgbe.rs:
```diff
@@ -0,0 +1,161 @@
+}
+
+/// Converts a `NetBufPtr` to `IxgbeNetBuf`.
+pub fn ixgbe_ptr_to_buf(ptr: NetBufPtr, pool: &Arc<MemPool>) -> DevResult<IxgbeNetBuf> {
```
no `pub`

equation314 commented on crates/driver_net/src/ixgbe.rs:
```diff
@@ -0,0 +1,161 @@
+
+impl From<ManuallyDrop<IxgbeNetBuf>> for NetBufPtr {
+    fn from(mut buf: ManuallyDrop<IxgbeNetBuf>) -> Self {
+        let buf_ptr = buf.packet_mut().as_mut_ptr();
```
Move `ManuallyDrop::new` here

equation314 commented on crates/driver_net/src/ixgbe.rs:
```diff
@@ -0,0 +1,161 @@
+            // RX queue is empty, receive from ixgbe NIC.
+            match self.inner.receive_packets(0, RECV_BATCH_SIZE, |rx_buf| {
+                let rx_buf = ManuallyDrop::new(rx_buf);
+                let rx_buf = rx_buf.into();
```
Use `NetBufPtr::from(rx_buf)` is more clearer.

equation314 commented on modules/axhal/src/platform/mod.rs:
```diff
@@ -10,7 +10,7 @@ cfg_if::cfg_if! {
         mod x86_pc;
         pub use self::x86_pc::*;
-    } else if #[cfg(all(target_arch = "riscv64", platform_family = "riscv64-qemu-virt"))] {
+    }else if #[cfg(all(target_arch = "riscv64", platform_family = "riscv64-qemu-virt"))] {
```
Cancel this change

equation314 commented on crates/driver_net/src/lib.rs:
```diff
@@ -3,7 +3,14 @@
+#![feature(box_into_inner)]
 
+#[macro_use]
+extern crate log;
```
unused `#[macro_use]` import

equation314 replied:
It' better to remove `extern crate log` here, and call `log::error!` in `ixgbe.rs`.

KuangjuX committed:
net: Add support for Intel 82599 (ixgbe) NIC & performed performance optimization.
- Referencing ixy.rs, implemented the Intel 82599 (ixgbe) driver and optimized its performance to a certain extent.
- Completed the detection and initialization of ixgbe in `axdriver`.
- Added a new platform `x86-pc-oslab` for adapting to physical machines and conducting real network card development and testing.
- Based on PR#99, re-implemented the ixgbe driver and protocol stack adaptation for the network interface.
- Currently, the performance tested using ab (Apache HTTP server benchmarking tool) can reach 25000+ requests/second. TCP packet receiving performance tested with iperf can reach 3.5G, while packet sending performance can reach 4.0G. UDP packet receiving performance tested can reach 3.4G (not accurately measured), and packet sending performance reaches 9+G. Using bw_client for testing, the packet receiving performance reaches at least 6G (not accurately measured, as the client did not fully send), and packet sending can fully saturate at 10G.
